ccvm_simulators/solvers/mf_solver.py

```diff
@@ -260,6 +315,84 @@ def tune(self, instances, post_processor, g=0.01):
         #       future consideration
         self.is_tuned = True
 
+    def optics_machine_energy(
+        self,
+        machine_parameters=None,
+    ):
+        """The wrapper function of calculating the maximum energy consumption of the
+        solver simulating on a MF-CCVM machine.
+
+        Args:
+            machine_parameters (dict, optional): Parameters of the. Defaults to None.
+
+        Raises:
+            ValueError: when the given machine parameters are not valid.
+            ValueError: when the given dataframe does not contain the required columns.
+
+        Returns:
+            Callable: A callable function that takes in a dataframe and problem size and
+                returns the maximum energy consumption of the solver.
+        """
+        if machine_parameters is None:
+            machine_parameters = self._default_optics_machine_parameters
+        else:
+            self._is_valid_optics_machine_parameters(machine_parameters)
+
+        def optics_machine_energy_callable(
+            matching_df: DataFrame,
+            problem_size: int,
+        ):
+            """Calculate the maximum power consumption of the solver simulating on a
+                MF-CCVM machine.
+
+            Args:
+                matching_df (DataFrame): The necessary data to calculate the maximum power.
+                problem_size (int): The size of the problem.
+
+            Raises:
+                ValueError: when the given dataframe does not contain the required columns.
+
+            Returns:
+                float: The maximum power consumption of the solver.
+            """
+            self._validate_dataframe_columns(matching_df)
+
+            try:
+                pump = self.parameter_key[problem_size]["pump"]
+                measure_strength = self.parameter_key[problem_size]["j"]
+            except KeyError as e:
+                raise KeyError(
+                    f"The parameter '{e.args[0]}' for the given instance size: {problem_size} is not defined."
+                ) from e
+
+            iterations = matching_df["iterations"].values
+            postprocessing_time = np.mean(matching_df["pp_time"].values)
+            roundtrip_time = (
+                (
+                    machine_parameters["FPGA_fixed"]
```

Comment:
@FarhadK-1QBit it looks like Estella correctly copied this from the Notion page, but I wasn't expecting MFSolver to have FPGA parameters. Is this an error, or am I misunderstanding it?

Reply:
Thanks for noting this, @KatieO4t1qbit ! This is actually correct. The MF-CCVM solver has an FPGA inside of it that performs parts of the computation. It's a hybrid optics-FPGA machine.